File: tvr-decoder/src/main/java/com/willcurrie/tlv/ConstructedBerTlv.java
package com.willcurrie.tlv;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class ConstructedBerTlv extends BerTlv {

    private final List<BerTlv> data;

    ConstructedBerTlv(Tag tag, List<BerTlv> data) {
        super(tag);
        this.data = data;
    }

    public byte[] getValue() {
        ByteArrayOutputStream value = new ByteArrayOutputStream();
        for (BerTlv tlv : data) {
            try {
				value.write(tlv.toBinary());
			} catch (IOException e) {
				throw new RuntimeException(e);
			}
        }
        return value.toByteArray();
    }

    public BerTlv findTlv(Tag tag) {
    	for (BerTlv tlv : data) {
    		if (Arrays.equals(tlv.getTag().getBytes(), tag.getBytes())) {
                return tlv;
            }
        }
        return null;
    }

    public List<BerTlv> findTlvs(Tag tag) {
        List<BerTlv> matches = new ArrayList<BerTlv>();
        for (BerTlv tlv : data) {
            if (Arrays.equals(tlv.getTag().getBytes(), tag.getBytes())) {
                matches.add(tlv);
            }
        }
        return matches;
    }

    public List<BerTlv> getChildren() {
        return data;
    }

    public String toString() {
        StringBuffer buffer = new StringBuffer("\nTag: ");
        buffer.append(getTag()).append('\n');
        for (BerTlv tlv : data) {
        	buffer.append(tlv).append('\n');
		}
        buffer.append("End tag: ").append(getTag()).append('\n');
        return buffer.toString();
    }
}

File: src/main/java/io/github/binaryfoo/tlv/ConstructedBerTlv.kt
package io.github.binaryfoo.tlv

import java.io.ByteArrayOutputStream
import java.io.IOException

/**
 * The V represents a set of TLVs.
 */
class ConstructedBerTlv(tag: Tag, private val children: List<BerTlv>) : BerTlv(tag) {

    override fun getChildren(): List<BerTlv> = children

    override fun getValue(): ByteArray {
        val value = ByteArrayOutputStream()
        for (tlv in children) {
            try {
                value.write(tlv.toBinary())
            } catch (e: IOException) {
                throw RuntimeException(e)
            }

        }
        return value.toByteArray()
    }

    override fun findTlv(tag: Tag): BerTlv? = children.firstOrNull { it.tag == tag }

    override fun findTlvs(tag: Tag): List<BerTlv> = children.filter { it.tag == tag }

    override fun toString(): String {
        val buffer = StringBuilder("\nTag: $tag\n")
        for (tlv in children) {
            buffer.append(tlv).append('\n')
        }
        buffer.append("End tag: $tag\n")
        return buffer.toString()
    }
}

